serving/serving.go

```diff
@@ -187,44 +194,38 @@ func (serv *FeatureServer) handleSplitDataRequest(
 		}
 
 		response := &pb.TrainingTestSplitResponse{
-			RequestType:       req.GetRequestType(),
+			RequestType:       ctx.req.GetRequestType(),
 			TrainingTestSplit: &pb.TrainingTestSplitResponse_Row{Row: sRow},
 		}
 
-		if err := stream.Send(response); err != nil {
-			logger.Errorw("Failed to write to stream", "Error", err)
+		if err := ctx.stream.Send(response); err != nil {
+			ctx.logger.Errorw("Failed to write to stream", "Error", err)
 			return err
 		}
 	} else {
-		// Once an iterator is finished we need to let the client know to stop iteration
-		// We do this so that the stream stays open so the client can still operate on the other iterator
-		serv.handleFinishedIterator(stream, req.GetRequestType(), isTestFinished, isTrainFinished, logger)
+		serv.handleFinishedIterator(ctx)
 	}
 
-	if err := thisIter.Err(); err != nil {
-		logger.Errorw("Dataset error", "Error", err)
-		return err
+	if thisIter.Err() != nil {
+		ctx.logger.Errorw("Dataset error", "Error", thisIter.Err())
+		return thisIter.Err()
 	}
 
 	return nil
 }
 
-func (serv *FeatureServer) handleFinishedIterator(
-	stream pb.Feature_TrainingTestSplitServer,
-	reqType pb.RequestType,
-	isTestFinished *bool,
-	isTrainFinished *bool,
-	logger *zap.SugaredLogger,
-) {
-	if reqType == pb.RequestType_TEST {
-		*isTestFinished = true
-	} else {
-		*isTrainFinished = true
+func (serv *FeatureServer) handleFinishedIterator(ctx *splitContext) {
+	if ctx.req.GetRequestType() == pb.RequestType_TEST {
+		*ctx.isTestFinished = true
+	} else if ctx.req.GetRequestType() == pb.RequestType_TRAINING {
+		*ctx.isTrainFinished = true
 	}
 
-	if *isTestFinished || *isTrainFinished {
-		if err := stream.Send(&pb.TrainingTestSplitResponse{IteratorDone: true}); err != nil {
-			logger.Errorw("Failed to write to stream", "Error", err)
+	if *ctx.isTestFinished && *ctx.isTrainFinished {
```

Comment:
I think theres a bug in this. When both streams are done, wont this execute the return before it sends the iterator done message?

Reply:
kinda, so there was a bug before where the stream would actually never close.

Iterator done is a signal for a single iterator being done, let me actually a comment for that. So if one of the iterators is done we send the iterator done singal so that the loop is stopped.

When both iterators are done, it returns and closes the stream, when the stream is a closed it automatically sends StopIteration message to the client)

There might be a better way to do all this